cpp/monitoring/gcm/exporter.cc

```diff
@@ -0,0 +1,264 @@
+              "Which GCE service account to use for pushing metrics.");
+
+namespace cert_trans {
+
+using std::chrono::seconds;
+using std::chrono::system_clock;
+using std::chrono::minutes;
+using std::make_pair;
+using std::mutex;
+using std::ostringstream;
+using std::placeholders::_1;
+using std::string;
+using std::thread;
+using std::unique_lock;
+using util::Executor;
+using util::SyncTask;
+
+
+namespace {
+
+
+// Label and Metric -name prefix.
+const char kCloudPrefix[] = "custom.cloudmonitoring.googleapis.com/ct/";
+
+
+inline std::string RFC3339Time(const system_clock::time_point& when) {
+  const std::time_t now_c(system_clock::to_time_t(when));
+  char buf[256];
+  CHECK(std::strftime(buf, sizeof(buf), "%FT%T.00Z", std::localtime(&now_c)));
+  return buf;
+}
+
+
+}  // namespace
+
+
+GCMExporter::GCMExporter(const string& instance_name, UrlFetcher* fetcher,
+                         Executor* executor)
+    : instance_name_(instance_name),
+      fetcher_(CHECK_NOTNULL(fetcher)),
+      executor_(CHECK_NOTNULL(executor)),
+      task_(executor_) {
+  RefreshCredentials();
+  CreateMetrics();
+  PushMetrics(nullptr);
+}
+
+
+GCMExporter::~GCMExporter() {
+  task_.task()->Cancel();
+  task_.Wait();
+}
+
+
+void GCMExporter::RefreshCredentials() {
+  VLOG(1) << "Refreshing GCM credentials...";
+  UrlFetcher::Request req(
+      (URL(FLAGS_google_compute_metadata_url + "/" +
+           FLAGS_google_compute_monitoring_service_account + "/token")));
+  req.headers.insert(make_pair("Metadata-Flavor", "Google"));
+
+  UrlFetcher::Response resp;
+  SyncTask task(executor_);
+  fetcher_->Fetch(req, &resp, task.task());
+  task.Wait();
+
+  // TODO(alcutter): Handle this better
+  CHECK_EQ(util::Status::OK, task.status());
+  CHECK_EQ(200, resp.status_code);
+  token_refreshed_at_ = system_clock::now();
+
+  JsonObject reply(resp.body);
+  CHECK(reply.Ok()) << "Failed to parse metadata JSON:\n" << resp.body;
+  JsonString bearer(reply, "access_token");
+  CHECK(bearer.Ok());
+  bearer_token_ = bearer.Value();
+
+  VLOG(1) << "GCM credentials refreshed";
+}
+
+
+namespace {
+
+
+void AddLabelDescription(const string& key, const string& desc,
+                         JsonArray* labels) {
+  JsonObject l;
+  l.Add("key", kCloudPrefix + key);
+  l.Add("description", desc);
+  CHECK_NOTNULL(labels)->Add(&l);
+}
+
+
+}  // namespace
+
+
+void GCMExporter::CreateMetrics() const {
+  const std::set<const Metric*> metrics(Registry::Instance()->GetMetrics());
+  for (auto& m : metrics) {
+    CHECK_NOTNULL(m);
+
+    // See
+    // https://cloud.google.com/monitoring/v2beta2/metricDescriptors#resource
+    // for a description of the structure we're building here.
+
+    JsonArray labels;
+    AddLabelDescription("instance",
+                        "Instance from which the sample originates.", &labels);
+    for (const auto& label : m->LabelNames()) {
+      AddLabelDescription(label, label, &labels);
```

Comment:
The `Metric` has a description (see L152), but there's no description of the `labels` other than its name (which already should really provide some insight into what it's there for.)